### DISCUSSION TITLE
Insert data into another table via filter on create event

### DISCUSSION THREAD

**Discussion by mayankpunetha007** [directus/directus#10311] [Archive]
Hi,

I have a Event table with a multi relation to a Multievent table.
On receiving an create/update on the the Event table I want to process a few things from an api and insert things into the multi table.

I created a hook in hooks/location/index.js which works fine apart from a few issues:

```
module.exports = function registerHook({ filter }, { services  }) {
    const { ItemsService } = services;
    filter('event.items.create',  async (input,meta, { database, schema, accountability }) => {
        const serviceSingle = new ItemsService('event', { database, schema, accountability });
        const serviceMulti = new ItemsService('event_multi', { database, schema, accountability });
        // Do some processing:
        // insert data
        serviceMulti.insert({'a':'1','b':'2'});
    });
};
```
2 questions here:
1. I get insert is not a function on serviceMulti, What is the rite way to insert to another table?
2. I am using a generated id for event How to link the event_multi to it, Generate my own id or is there a way to receive data after directus has generated an id?

And an unrelated question to the above 2:
Is there a easy way to debug directus code like starting in debug mode?

Currently I use:
```npx directus start```

In order to start the application.

**Comment by joselcvarela** [comment]
Hello @mayankpunetha007 
1. The proper method to insert things into database is `createOne` or `createMany` so in your case it should be `serviceMulti.createOne({'a':'1','b':'2'});`
Also keep in mind this method is asynchronous so you need to add `await` before calling it, like `await serviceMulti.createOne({'a':'1','b':'2'});`
2. Right now you are creating things in another table before they were created on the main table. To have access to generated IDs you will need to do this operation after the items creation. Then, you can have access to the generated ID via first argument property `key`. In order to do so, you should use `action` instead of `filter`. In your case 
```js
action('event.items.create',  async ({ payload, key }, { database, schema, accountability }) => {
```
`payload` should have the same information as your `input` but could have also extra fields for database

And yes it is possible to debug Directus You can do it for any Node.js cli command. Here is how:
```sh
$ cd your-directus-project
node --inspect node_modules/.bin/directus start
```

Then in your debugger client, like VSCode, create a configuration to attach debugger into 9229 port like: 
```json
        {
            "name": "Attach",
            "port": 9229,
            "restart": true,
            "request": "attach",
            "skipFiles": [
                "<node_internals>/**"
            ],
            "type": "pwa-node"
        }
```

**Reply by mayankpunetha007** [reply]
Thanks a lot for taking the time to respond.

1 and 2 are resolved, Thanks for the help! :)

However on the debug side of things I am getting the following error on trying to debug:
```
node --inspect node_modules/.bin/directus start
Debugger listening on ws://127.0.0.1:9229/ef385058-b59b-42f8-a21e-4bb8a06211be
For help, see: https://nodejs.org/en/docs/inspector
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^
SyntaxError: missing ) after argument list
←[90m    at wrapSafe (internal/modules/cjs/loader.js:979:16)←[39m
←[90m    at Module._compile (internal/modules/cjs/loader.js:1027:27)←[39m
←[90m    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)←[39m
←[90m    at Module.load (internal/modules/cjs/loader.js:928:32)←[39m
←[90m    at Function.Module._load (internal/modules/cjs/loader.js:769:14)←[39m
←[90m    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)←[39m
←[90m    at internal/main/run_main_module.js:17:47←[39m
```

Any idea what might be causing it?

I am on windows 10 node version v14.15.4 on an exFat partition.

Thanks!

**Reply by joselcvarela** [reply]
Do you have bash on Windows like the installed by Git or with WSL?
Try to run from there. It should work in theory 👍

**Reply by mayankpunetha007** [reply]
Yes was trying for gitbash.

But something seems wrong in getting the basedir in the directus bash file

The following worked however, What the bash was doing
node --inspect node_modules/directus/cli.js start

Thanks again for the help! :)

**Reply by mayankpunetha007** [reply]
@joselcvarela : Sorry a follow-up question on this.

Do filters work with directus collections as well?
For example I have added a few columns to the directus_files.item.create

And I am trying to update them as follows:


```
    filter('directus_files.items.create',  async (input, meta, { database, schema, accountability }) => {
        input['url'] = storageUrl+input['filename_download'];
        return input;
    });
```

Is this something supported?

**Reply by joselcvarela** [reply]
Hello again @mayankpunetha007
It is not possible to use `filter` with file uploads because you are handling stream with multiple chunks, so the upload needs to complete in order to acces the file. Because of this you should use `action('files.upload')` instead https://docs.directus.io/extensions/hooks/#action-events

**Reply by mayankpunetha007** [reply]
Thank you! :)
